File: app/src/main/java/wangdaye/com/geometricweather/common/ui/widgets/TagView.java
package wangdaye.com.geometricweather.common.ui.widgets;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Outline;
import android.graphics.Paint;
import android.graphics.RectF;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewOutlineProvider;

import androidx.annotation.ColorInt;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatTextView;

import wangdaye.com.geometricweather.R;

public class TagView extends AppCompatTextView {

    private final RectF mOutline;

    private final Paint mPaint;

    private boolean mChecked;

    private @ColorInt int mCheckedBackgroundColor;
    private @ColorInt int mUncheckedBackgroundColor;

    public TagView(Context context) {
        this(context, null);
    }

    public TagView(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public TagView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        mOutline = new RectF();

        mPaint = new Paint();
        mPaint.setAntiAlias(true);
        mPaint.setStyle(Paint.Style.FILL);

        TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.TagView, defStyleAttr, 0);
        setChecked(a.getBoolean(R.styleable.TagView_checked, false));
        setCheckedBackgroundColor(a.getColor(R.styleable.TagView_checked_background_color, Color.WHITE));
        setUncheckedBackgroundColor(a.getColor(R.styleable.TagView_unchecked_background_color, Color.LTGRAY));
        a.recycle();
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);

        mOutline.set(0, 0, getMeasuredWidth(), getMeasuredHeight());

        setClipToOutline(true);
        setOutlineProvider(new ViewOutlineProvider() {
            @Override
            public void getOutline(View view, Outline viewOutline) {
                viewOutline.setRoundRect(
                        (int) mOutline.left,
                        (int) mOutline.top,
                        (int) mOutline.right,
                        (int) mOutline.bottom,
                        mOutline.height() / 2
                );
            }
        });
    }

    @Override
    protected void onDraw(Canvas canvas) {
        mPaint.setColor(mChecked ? mCheckedBackgroundColor : mUncheckedBackgroundColor);
        canvas.drawRoundRect(
                mOutline, mOutline.height() / 2, mOutline.height() / 2, mPaint);

        super.onDraw(canvas);
    }

    public final boolean isChecked() {
        return mChecked;
    }

    public void setChecked(boolean checked) {
        mChecked = checked;
        invalidate();
    }

    public int getCheckedBackgroundColor() {
        return mCheckedBackgroundColor;
    }

    public void setCheckedBackgroundColor(int checkedBackgroundColor) {
        mCheckedBackgroundColor = checkedBackgroundColor;
        invalidate();
    }

    public int getUncheckedBackgroundColor() {
        return mUncheckedBackgroundColor;
    }

    public void setUncheckedBackgroundColor(int uncheckedBackgroundColor) {
        mUncheckedBackgroundColor = uncheckedBackgroundColor;
        invalidate();
    }
}

File: app/src/main/java/org/breezyweather/ui/common/widgets/TagView.kt
/**
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.ui.common.widgets

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Outline
import android.graphics.Paint
import android.graphics.RectF
import android.util.AttributeSet
import android.view.View
import android.view.ViewOutlineProvider
import android.view.accessibility.AccessibilityNodeInfo
import androidx.annotation.ColorInt
import androidx.appcompat.widget.AppCompatTextView
import androidx.core.content.withStyledAttributes
import androidx.core.view.AccessibilityDelegateCompat
import androidx.core.view.ViewCompat
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat
import org.breezyweather.R

class TagView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0,
) : AppCompatTextView(context, attrs, defStyleAttr) {
    private val mOutline: RectF = RectF()
    private val mPaint: Paint = Paint().apply {
        isAntiAlias = true
        style = Paint.Style.FILL
    }
    private var mChecked = false

    @ColorInt
    private var mCheckedBackgroundColor = 0

    @ColorInt
    private var mUncheckedBackgroundColor = 0

    init {
        context.withStyledAttributes(attrs, R.styleable.TagView, defStyleAttr, 0) {
            mChecked = getBoolean(R.styleable.TagView_checked, false)
            mCheckedBackgroundColor =
                getColor(R.styleable.TagView_checked_background_color, Color.WHITE)
            mUncheckedBackgroundColor =
                getColor(R.styleable.TagView_unchecked_background_color, Color.LTGRAY)
        }

        ViewCompat.setAccessibilityDelegate(
            this,
            object : AccessibilityDelegateCompat() {
                override fun onInitializeAccessibilityNodeInfo(
                    host: View,
                    info: AccessibilityNodeInfoCompat,
                ) {
                    super.onInitializeAccessibilityNodeInfo(host, info)
                    info.roleDescription = context.resources.getString(androidx.compose.ui.R.string.tab)
                }
            }
        )
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        mOutline.set(0f, 0f, measuredWidth.toFloat(), measuredHeight.toFloat())
        clipToOutline = true
        outlineProvider = object : ViewOutlineProvider() {
            override fun getOutline(view: View, viewOutline: Outline) {
                viewOutline.setRoundRect(
                    mOutline.left.toInt(),
                    mOutline.top.toInt(),
                    mOutline.right.toInt(),
                    mOutline.bottom.toInt(),
                    mOutline.height() / 2
                )
            }
        }
    }

    override fun onDraw(canvas: Canvas) {
        mPaint.color = if (mChecked) mCheckedBackgroundColor else mUncheckedBackgroundColor
        canvas.drawRoundRect(mOutline, mOutline.height() / 2, mOutline.height() / 2, mPaint)
        super.onDraw(canvas)
    }

    var isChecked: Boolean
        get() = mChecked
        set(checked) {
            mChecked = checked
            invalidate()
        }
    var checkedBackgroundColor: Int
        get() = mCheckedBackgroundColor
        set(checkedBackgroundColor) {
            mCheckedBackgroundColor = checkedBackgroundColor
            invalidate()
        }
    var uncheckedBackgroundColor: Int
        get() = mUncheckedBackgroundColor
        set(uncheckedBackgroundColor) {
            mUncheckedBackgroundColor = uncheckedBackgroundColor
            invalidate()
        }

    override fun onInitializeAccessibilityNodeInfo(info: AccessibilityNodeInfo) {
        super.onInitializeAccessibilityNodeInfo(info)
        info.isSelected = isChecked
        info.isCheckable = false
        info.isClickable = !isChecked // TODO: Not working
    }
}
